pkg/backend/identityapi/register_client.go

```diff
@@ -42,34 +61,121 @@ type RegisterEntity struct {
+) (RegisterClient, error) {
 	if compressionLevel < gzip.NoCompression || compressionLevel > gzip.BestCompression {
 		return nil, fmt.Errorf("gzip: invalid compression level: %d", compressionLevel)
 	}
+	icfg := identity.NewConfiguration()
+	icfg.BasePath = svcUrl + identityPath
+	icfg.Debug = true
+	// TODO: add the global HTTP client here
+	// icfg.HTTPClient = httpClient
+	identityClient := identity.NewAPIClient(icfg)
 	return &registerClient{
 		svcUrl:           strings.TrimSuffix(svcUrl, "/"),
 		licenseKey:       licenseKey,
 		userAgent:        userAgent,
 		httpClient:       httpClient,
 		compressionLevel: compressionLevel,
+		apiClient:        identityClient.DefaultApi,
 	}, nil
 }
 
+func (rc *registerClient) RegisterBatchEntities(agentEntityID entity.ID, entities []protocol.Entity) (resp RegisterBatchEntityResponse, duration time.Duration, err error) {
+
+	ctx := context.Background()
+
+	registerRequests := make([]identity.RegisterRequest, len(entities))
+
+	for i := range entities {
+		registerRequests[i] = newRegisterRequest(entities[i])
+	}
+
+	localVarOptionals := &identity.RegisterBatchPostOpts{
+		XNRIAgentEntityId: optional.NewInt64(int64(agentEntityID)),
+	}
+
+	apiReps, httpResp, err := rc.apiClient.RegisterBatchPost(ctx, rc.userAgent, rc.licenseKey, registerRequests, localVarOptionals)
+	if err != nil {
+		rlog.
+			WithError(err).
+			WithField("XNRIAgentEntityId", agentEntityID).
+			WithField("status", httpResp.StatusCode).
+			WithField("RegisterRequests", registerRequests).
+			Debug("Failed making a Register Batch Post request")
+		return resp, time.Second, err // TODO add right duration
+	}
+
+	resp = make(RegisterBatchEntityResponse, len(apiReps))
+
+	for i := range apiReps {
+		resp[i] = RegisterEntityResponse{
+			ID:   entity.ID(apiReps[i].EntityId),
+			Key:  entity.Key(apiReps[i].EntityName),
+			Name: apiReps[i].EntityName,
+		}
+	}
+
+	return resp, time.Second, err
+}
+
+func newRegisterRequest(entity protocol.Entity) identity.RegisterRequest {
+	registerRequest := identity.RegisterRequest{
+		EntityType:  entity.Type,
+		EntityName:  entity.Name,
+		DisplayName: entity.DisplayName,
+		Metadata:    convertMetadataToMapStringString(entity.Metadata),
+	}
+	return registerRequest
+}
+
+func (rc *registerClient) RegisterEntity(agentEntityID entity.ID, ent protocol.Entity) (resp RegisterEntityResponse, err error) {
+
+	ctx := context.Background()
+	registerRequest := newRegisterRequest(ent)
+	localVarOptionals := &identity.RegisterPostOpts{
+		XNRIAgentEntityId: optional.NewInt64(int64(agentEntityID)),
+	}
+
+	apiReps, httpResp, err := rc.apiClient.RegisterPost(ctx, rc.userAgent, rc.licenseKey, registerRequest, localVarOptionals)
+	if err != nil {
+		rlog.
+			WithError(err).
+			WithField("XNRIAgentEntityId", agentEntityID).
+			WithField("status", httpResp.StatusCode).
+			WithField("RegisterRequest", registerRequest).
+			Debug("Failed making a Register Post request")
+		return resp, err
```

Comment:
not a fan of logging an returning the same error

I think callers should be able to handle the returned error, logging only when required

only one of these fields is not available for the caller "StatusCode", I guess we can return this one within the error or response object if required